packages/extension-dapp/src/compat/web3Source.ts

```diff
@@ -90,7 +90,12 @@ function injectWeb3(win: Web3Window): void {
             },
           },
           signer: {
+            signPayload:async (payload: SignerPayloadJSON) : Promise<SignerResult>=>{
+              console.log('signPayload')
+              return {id:0,signature:await win.web3.eth.sign(JSON.stringify(payload), payload.address)}
```

Comment:
```suggestion
              const raw = regsirty.createPayload('ExtrinsicPayload', payload).toU8a();
              
              return {id:0,signature:await win.web3.eth.sign(raw), payload.address)}
```

Pass the raw data, exactly as to be signed - not sure about `registry` here, not sure what is available in this context.

Reply:
PS; Sam with raw, you probably only need Raw - Metamask is never going to add decoding, so the logic in the API is -

- if signPayload, use that
- if sigRaw, use that
- otherwise, bail

So raw ill just be the actual raw data and cleans it up nicely.

Reply:
Im not sure I understand what I should do for the signRaw function

Reply:
It seems fine as-is. Remove `signPayload`, only keep that one.

Reply:
but it seems signPayload is the one being called for transfers

Reply:
If available, like I mentioned, the order is -

- use signPayload, if available, other ...
- use signRaw, if available, otherwise ..
- ... bail

https://github.com/polkadot-js/api/blob/master/packages/api/src/submittable/createClass.ts#L292-L298

Reply:
ok and should I sign raw.data, registry.createType('ExtrinsicPayload', raw) or registry.createType('ExtrinsicPayload', raw.data);

Reply:
Just the `raw.data` in `signRaw` is 100% - it is the actual data.

Reply:
I updated back to using only signRaw and it gets stuck at the 'signing' step that I screenshot below. Let me know if it works for you

Reply:
Does this get logged?

https://github.com/polkadot-js/extension/pull/566/files#diff-85bb08733e1b7975e81532efecc738a9900827f7c54c78069e49d4c0de1a48e8R77

Just display the address and payload there as well - worried that is is an apps-specific issue.

Reply:
Yes it does get log

```
console.log('signraw',raw.data,raw.address)
            console.log('signature', await win.web3.eth.sign(raw.data, raw.address));
```
logs this
<img width="526" alt="Screenshot 2020-12-10 at 17 27 07" src="https://user-images.githubusercontent.com/16758383/101799907-f3bf9f00-3b0c-11eb-9970-4636f469385c.png">

Reply:
so it seems the signature is never resolved by MetaMask

Reply:
Ok, that data looks correct. So it seems MM doesn't come back. May be useful having a try/catch around the `await` to see if anything gets thrown. It may not like raw hex.

Reply:
good call. It doesnt log anything unfortunately. just realized that web3.eth.sign is sync. But wrapping the callback in a promise doesn't return anything either. Might be a problem with Metamask... They have been warning against eth_sign for a while now because it is a security issue

Reply:
also it looks like the message would be appended with "\x19Ethereum Signed Message:\n" https://eth.wiki/json-rpc/API#eth_sign

Reply:
Ahhh, forgot about the below, even though it is something in tests at least in 2 places. 

That would be an issue.

Reply:
Yes indeed. Another approach would be use the bi-compatibility of Moonbeam to post an EVM compatible tx through subtrate. Let me check with the team and get back to you